In [1]:
import numpy as np
import pandas as pd

In [2]:
movies_df=pd.read_csv("tmdb/tmdb_movies_data.csv")
movies_df.head()

id    imdb_id  popularity     budget     revenue  \
0  135397  tt0369610   32.985763  150000000  1513528810   
1   76341  tt1392190   28.419936  150000000   378436354   
2  262500  tt2908446   13.112507  110000000   295238201   
3  140607  tt2488496   11.173104  200000000  2068178225   
4  168259  tt2820852    9.335014  190000000  1506249360   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                            homepage          director  \
0                      http://www.jurassicworld.com/   Colin Trevorrow   
1                        http://www.madmaxmovie.com/     George Miller   
2     http://www.thedivergentseries.movie/#insurgent  Robert Schwentke   
3  http://www.starwars.com/films/star-wars-episod...       J.J. Abrams   
4                           http://www.furious7.com/         James Wan   

                         tagline  ...  \
0              The park is open.  ...   
1             What a Lovely Day.  ...   
2     One Choice Can Destroy You  ...   
3  Every generation has a story.  ...   
4            Vengeance Hits Home  ...   

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   
2  Beatrice Prior must confront her inner demons ...     119   
3  Thirty years after defeating the Galactic Empi...     136   
4  Deckard Shaw seeks revenge against Dominic Tor...     137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...     6/9/2015       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...    5/13/2015       6185   
2  Summit Entertainment|Mandeville Films|Red Wago...    3/18/2015       2480   
3          Lucasfilm|Truenorth Productions|Bad Robot   12/15/2015       5292   
4  Universal Pictures|Original Film|Media Rights ...     4/1/2015       2947   

   vote_average  release_year   budget_adj   revenue_adj  
0           6.5          2015  137999939.3  1.392446e+09  
1           7.1          2015  137999939.3  3.481613e+08  
2           6.3          2015  101199955.5  2.716190e+08  
3           7.5          2015  183999919.0  1.902723e+09  
4           7.3          2015  174799923.1  1.385749e+09  

[5 rows x 21 columns]

Taking only the columns needed for content based recommendation in another dataframe.

In [3]:
movies_tmdb=movies_df.loc[:,["original_title","cast","genres"]]

In [4]:
movies_tmdb.head()

original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                      genres  
0  Action|Adventure|Science Fiction|Thriller  
1  Action|Adventure|Science Fiction|Thriller  
2         Adventure|Science Fiction|Thriller  
3   Action|Adventure|Science Fiction|Fantasy  
4                      Action|Crime|Thriller

In [5]:
movies_tmdb.shape

(10866, 3)

In [6]:
movies_tmdb.isna().sum()

original_title     0
cast              76
genres            23
dtype: int64

Dropping the null values before performing any data processing operations

In [7]:
movies_tmdb.dropna(inplace=True)

In [8]:
movies_tmdb.isna().sum()

original_title    0
cast              0
genres            0
dtype: int64

Replacing the separator with space, as to get each word separately.

In [9]:
movies_tmdb["cast"] = movies_tmdb["cast"].apply(lambda x: x.replace("|"," "))
movies_tmdb["genres"] = movies_tmdb["genres"].apply(lambda x: x.replace("|"," "))

In [10]:
movies_tmdb.head()

original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt Bryce Dallas Howard Irrfan Khan Vi...   
1  Tom Hardy Charlize Theron Hugh Keays-Byrne Nic...   
2  Shailene Woodley Theo James Kate Winslet Ansel...   
3  Harrison Ford Mark Hamill Carrie Fisher Adam D...   
4  Vin Diesel Paul Walker Jason Statham Michelle ...   

                                      genres  
0  Action Adventure Science Fiction Thriller  
1  Action Adventure Science Fiction Thriller  
2         Adventure Science Fiction Thriller  
3   Action Adventure Science Fiction Fantasy  
4                      Action Crime Thriller

Making a column named combined, which has all the keywords of cast and genres of each movie. 

In [32]:
movies_tmdb["combined"]=movies_tmdb["genres"] + " " + movies_tmdb["cast"]
movies_tmdb.head()

title  \
0                jurassic world   
1            mad max: fury road   
2                     insurgent   
3  star wars: the force awakens   
4                     furious 7   

                                                cast  \
0  Chris Pratt Bryce Dallas Howard Irrfan Khan Vi...   
1  Tom Hardy Charlize Theron Hugh Keays-Byrne Nic...   
2  Shailene Woodley Theo James Kate Winslet Ansel...   
3  Harrison Ford Mark Hamill Carrie Fisher Adam D...   
4  Vin Diesel Paul Walker Jason Statham Michelle ...   

                                      genres  \
0  Action Adventure Science Fiction Thriller   
1  Action Adventure Science Fiction Thriller   
2         Adventure Science Fiction Thriller   
3   Action Adventure Science Fiction Fantasy   
4                      Action Crime Thriller   

                                            combined  
0  Action Adventure Science Fiction Thriller Chri...  
1  Action Adventure Science Fiction Thriller Tom ...  
2  Adventure Science Fiction Thriller Shailene Wo...  
3  Action Adventure Science Fiction Fantasy Harri...  
4  Action Crime Thriller Vin Diesel Paul Walker J...

Checking if all the words are spaced properly and no abnormality is present.

In [12]:
movies_tmdb["combined"][0]

"Action Adventure Science Fiction Thriller Chris Pratt Bryce Dallas Howard Irrfan Khan Vincent D'Onofrio Nick Robinson"

Renaming "original_title" to "title" for simplicity.

In [13]:
movies_tmdb.rename(columns={"original_title":"title"},inplace=True)
movies_tmdb.head()

title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt Bryce Dallas Howard Irrfan Khan Vi...   
1  Tom Hardy Charlize Theron Hugh Keays-Byrne Nic...   
2  Shailene Woodley Theo James Kate Winslet Ansel...   
3  Harrison Ford Mark Hamill Carrie Fisher Adam D...   
4  Vin Diesel Paul Walker Jason Statham Michelle ...   

                                      genres  \
0  Action Adventure Science Fiction Thriller   
1  Action Adventure Science Fiction Thriller   
2         Adventure Science Fiction Thriller   
3   Action Adventure Science Fiction Fantasy   
4                      Action Crime Thriller   

                                            combined  
0  Action Adventure Science Fiction Thriller Chri...  
1  Action Adventure Science Fiction Thriller Tom ...  
2  Adventure Science Fiction Thriller Shailene Wo...  
3  Action Adventure Science Fiction Fantasy Harri...  
4  Action Crime Thriller Vin Diesel Paul Walker J...

Converting all the movie titles to lowercase for search simplicity while recommending.

In [14]:
movies_tmdb["title"] = movies_tmdb["title"].str.lower()
movies_tmdb.head()

title  \
0                jurassic world   
1            mad max: fury road   
2                     insurgent   
3  star wars: the force awakens   
4                     furious 7   

                                                cast  \
0  Chris Pratt Bryce Dallas Howard Irrfan Khan Vi...   
1  Tom Hardy Charlize Theron Hugh Keays-Byrne Nic...   
2  Shailene Woodley Theo James Kate Winslet Ansel...   
3  Harrison Ford Mark Hamill Carrie Fisher Adam D...   
4  Vin Diesel Paul Walker Jason Statham Michelle ...   

                                      genres  \
0  Action Adventure Science Fiction Thriller   
1  Action Adventure Science Fiction Thriller   
2         Adventure Science Fiction Thriller   
3   Action Adventure Science Fiction Fantasy   
4                      Action Crime Thriller   

                                            combined  
0  Action Adventure Science Fiction Thriller Chri...  
1  Action Adventure Science Fiction Thriller Tom ...  
2  Adventure Science Fiction Thriller Shailene Wo...  
3  Action Adventure Science Fiction Fantasy Harri...  
4  Action Crime Thriller Vin Diesel Paul Walker J...

In [15]:
movies_tmdb.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10768 entries, 0 to 10865
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   title     10768 non-null  object
 1   cast      10768 non-null  object
 2   genres    10768 non-null  object
 3   combined  10768 non-null  object
dtypes: object(4)
memory usage: 678.7+ KB


# Most Important Step
Here we reset the index of movies_tmdb as to avoid the index blunder made by dropping the na values(You can see above Int64Index: 10768 entries, 0 to 10865 which should be RangeIndex: 10768 entries, 0 to 10767).
Due to which, while making the Similarity Matrix the indexes get misinterpreted. 

In [16]:
movies_tmdb.reset_index(drop=True,inplace=True)

In [17]:
movies_tmdb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10768 entries, 0 to 10767
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   title     10768 non-null  object
 1   cast      10768 non-null  object
 2   genres    10768 non-null  object
 3   combined  10768 non-null  object
dtypes: object(4)
memory usage: 336.6+ KB


Importing the necessary modules from sklearn library.

In [18]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity



Making an instance of Count Vectorizer and applying fit_transform on combined column of movies_tmdb dataframe which will return the frequency of feature names for each movie.

In [19]:
cv = CountVectorizer()
count_vectorizer_matrix = cv.fit_transform(movies_tmdb["combined"])
count_vectorizer_matrix

<10768x15906 sparse matrix of type '<class 'numpy.int64'>'
	with 136148 stored elements in Compressed Sparse Row format>

In [20]:
cv.get_feature_names()

['50',
 'aalderks',
 'aaliyah',
 'aames',
 'aamir',
 'aanjjan',
 'aaran',
 'aarif',
 'aaron',
 'aarti',
 'aasha',
 'aasif',
 'aattou',
 'abascal',
 'abbass',
 'abbes',
 'abbey',
 'abbiati',
 'abbie',
 'abbott',
 'abby',
 'abdalla',
 'abdelhafid',
 'abdi',
 'abdoo',
 'abdul',
 'abe',
 'abebrese',
 'abeckaser',
 'abel',
 'abelanski',
 'abell',
 'abercrombie',
 'abercromby',
 'abhin',
 'abhishek',
 'abi',
 'abian',
 'abigail',
 'abiteboul',
 'abittan',
 'abkarian',
 'abke',
 'able',
 'abo',
 'abossolo',
 'abraham',
 'abrahams',
 'abramoviä',
 'abrams',
 'abramsohn',
 'abreu',
 'abromeit',
 'abry',
 'absera',
 'absolom',
 'ac',
 'accorsi',
 'ace',
 'acharanat',
 'acheche',
 'acheson',
 'acker',
 'ackerman',
 'ackermann',
 'ackland',
 'acklen',
 'ackles',
 'ackman',
 'acosta',
 'acovone',
 'acqua',
 'acres',
 'action',
 'acuã',
 'ad',
 'adachi',
 'adam',
 'adamaly',
 'adamczyk',
 'adamek',
 'adames',
 'adamo',
 'adams',
 'adamson',
 'adamthwaite',
 'adan',
 'adar',
 'adcock',
 'adcox',
 'ad

In [33]:
count_vectorizer_matrix.toarray()[1]

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

Making a similarity matrix which holds the similarity score for each movie with every other movie.

In [23]:
similarity_mat=cosine_similarity(count_vectorizer_matrix)

In [24]:
similarity_mat

array([[1.        , 0.3125    , 0.26726124, ..., 0.        , 0.07216878,
        0.        ],
       [0.3125    , 1.        , 0.26726124, ..., 0.        , 0.07216878,
        0.07537784],
       [0.26726124, 0.26726124, 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.08333333,
        0.        ],
       [0.07216878, 0.07216878, 0.        , ..., 0.08333333, 1.        ,
        0.08703883],
       [0.        , 0.07537784, 0.        , ..., 0.        , 0.08703883,
        1.        ]])

Taking a sample movie for trial

In [25]:
movie_index = movies_tmdb.loc[movies_tmdb['title']=="furious 7"].index[0]

Making a list of score of similar movies to the sample movie

In [26]:
movie_list = list(enumerate(similarity_mat[movie_index]))

Sorting the movies in descending order, as higher the similarity score more similar is the movie w.r.t sample movie.

In [27]:
movie_list = sorted(movie_list , key = lambda x:x[1] ,reverse=True)

Dropping the first movie in the list, as it is the same movie(sample movie)

In [28]:
movie_list = movie_list[1:11]

In [30]:
similarmovies = []
for i in range(len(movie_list)):
    a = movie_list[i][0]
    similarmovies.append(movies_tmdb['title'][a])

As we can see here, our recommendation system has returned a meaningful output for our sample movie "furious 7" as the movies with the most similar content would be its sequel or prequel and then would be movies with same actors, directors or genres.

In [31]:
similarmovies

['the fast and the furious',
 'fast five',
 'fast & furious 6',
 'fast & furious',
 'the death and life of bobby z',
 'turbo charged prelude to 2 fast 2 furious',
 'los bandoleros',
 '2 fast 2 furious',
 'the mechanic',
 'blitz']